packages/graphql/src/utils/join-predicates.ts

```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright (c) "Neo4j"
+ * Neo4j Sweden AB [http://neo4j.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+export default function joinPredicates(predicates: string[], joinType: "AND" | "OR"): string {
+    if (predicates.length === 0) {
+        return "";
+    }
+
+    if (predicates.length === 1) {
+        return predicates[0];
+    }
+
+    return `(${
+        joinType === "AND" ? predicates.join(" AND ") : predicates.map((predicate) => `(${predicate})`).join(" OR ")
```

Comment:
I'm not sure I understand the reasoning behind treating AND and OR differently. We could wrap all predicated with parenthesis to improve consistency, making this function much simpler:

```ts
export default function joinPredicates(predicates: string[], joinType: "AND" | "OR"): string {
    if (predicates.length === 0) {
        return "";
    }

    return `(${
        predicates.map(predicate=>`(${predicate})`.join(` ${joinType} `);
     })`;
  }
```

This may lead to unnecessary parenthesis, but should not be a problem. Thoughts on this?.
This is not too important of a change so don't worry about it if it requires too many changes in the tests

Reply:
This is how I originally planned to implement the change, I agree that it's better to be explicit with parentheses both for OR and for AND.

The code generates a lot of ANDs and not too many ORs, and the unit tests therefore largely contain tests which result in ANDs.

When I implemented the initial design to wrap all predicates in additional parentheses, several dozen unit tests were failing which would have taken a lot of time to go through and fix, and created an incredibly noisy PR.  In the majority of cases, where only ANDs were being generated, there was no issue as the order of evaluation wouldn't matter; it's only when operators of different precedence become mixed it's a problem.

As a consequence, to minimize noise (and hopefully get this fix accepted ASAP, I decided to target the issue around ORs specifically.

I think it would be a good idea to revisit later, but it's not a change I want to make right now.